# LAPORAN PRAKTIKUM KECERDASAN ARTIFISIAL LANJUT
**Nama: Sayyidah Fatimah Azzahra**

**NIM: 235150200111064**

**Kelas: TIF-E**

---
## Bab 4 : Pemrosesan Awal Data


### Soal 1
**1. Implementasikan metode normalisasi Z-score dengan cara membuat fungsi bernama score.**

In [33]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

data = pd.read_csv('dataset/Iris.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    object 
dtypes: float64(4), int64(1), object(1)
memory usage: 7.2+ KB


In [34]:
data = data.drop(columns=['Id'])

In [35]:
def zscore(data):
    return (data - data.mean()) / data.std()

In [36]:
fitur_numerik = data.select_dtypes(include=['int64', 'float64']).columns

normalized_data = data[fitur_numerik].apply(zscore)
print(normalized_data)
print(data)

     SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
0        -0.897674      1.028611      -1.336794     -1.308593
1        -1.139200     -0.124540      -1.336794     -1.308593
2        -1.380727      0.336720      -1.393470     -1.308593
3        -1.501490      0.106090      -1.280118     -1.308593
4        -1.018437      1.259242      -1.336794     -1.308593
..             ...           ...            ...           ...
145       1.034539     -0.124540       0.816888      1.443121
146       0.551486     -1.277692       0.703536      0.918985
147       0.793012     -0.124540       0.816888      1.050019
148       0.430722      0.797981       0.930239      1.443121
149       0.068433     -0.124540       0.760212      0.787951

[150 rows x 4 columns]
     SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm         Species
0              5.1           3.5            1.4           0.2     Iris-setosa
1              4.9           3.0            1.4           0.2     Iris-setos

#### Penjelasan

Setiap nilai pada kolom-kolom pada dataset dinormalisasi dengan menggunakan fungsi zscore, yang bekerja dengan cara mengurangi nilai dengan rata-rata dan membaginya dengan standar deviasi.

### Soal 2

**2. Normalisasikan DataFrame data_imputasi menggunakan fungsi zscore. Simpan hasilnya pada
DataFrame bernama data_zscore.**

In [37]:
def imputasi(df_input) :
    list_columns = df_input.columns #mendapatkan daftar kolom pada DataFrame
    class_column = list_columns[-1] #kolom terakhir merupakan kolom kelas
    for column in list_columns [:-1]:
        # Penggantian nilai NaN dilakukan per kolom
        df_input[column] = df_input[column].fillna(
            df_input.groupby(class_column)[column].transform('mean'))
    return df_input

data_imputasi = imputasi(data)
data_imputasi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   SepalLengthCm  150 non-null    float64
 1   SepalWidthCm   150 non-null    float64
 2   PetalLengthCm  150 non-null    float64
 3   PetalWidthCm   150 non-null    float64
 4   Species        150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [38]:
data_zscore = data_imputasi[fitur_numerik].apply(zscore)
data_zscore

,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm
0,-0.897674,1.028611,-1.336794,-1.308593
1,-1.139200,-0.124540,-1.336794,-1.308593
2,-1.380727,0.336720,-1.393470,-1.308593
3,-1.501490,0.106090,-1.280118,-1.308593
4,-1.018437,1.259242,-1.336794,-1.308593
...,...,...,...,...
145,1.034539,-0.124540,0.816888,1.443121
146,0.551486,-1.277692,0.703536,0.918985
147,0.793012,-0.124540,0.816888,1.050019
148,0.430722,0.797981,0.930239,1.443121


#### Penjelasan

Sama halnya seperti soal sebelumnya, saya mengimplementasikan fungsi zcore ini pada dataset, namun setelah dilakukan imputasi dengan mean.

### Soal 3

**3. Jelaskan perbedaan hasil normalisasi MinMax dan Z-Score. Petunjuk: cek nilai rentang, rata-rata dan simpangan baku dari data_zscore.**

In [39]:
def minmax_normalization(data):
    min_val = min(data)
    max_val = max(data)
    normalized_data = [(x - min_val) / (max_val - min_val) for x in data]
    return normalized_data

data_minmax = data_imputasi[fitur_numerik].apply(minmax_normalization)
data_minmax.head()

,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm
0,0.222222,0.625000,0.067797,0.041667
1,0.166667,0.416667,0.067797,0.041667
2,0.111111,0.500000,0.050847,0.041667
3,0.083333,0.458333,0.084746,0.041667
4,0.194444,0.666667,0.067797,0.041667


In [40]:
print(data_zscore.describe())
print(data_minmax.describe())

       SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
count   1.500000e+02  1.500000e+02   1.500000e+02  1.500000e+02
mean   -5.684342e-16 -6.158037e-16   1.894781e-16 -1.894781e-16
std     1.000000e+00  1.000000e+00   1.000000e+00  1.000000e+00
min    -1.863780e+00 -2.430844e+00  -1.563497e+00 -1.439627e+00
25%    -8.976739e-01 -5.858010e-01  -1.223442e+00 -1.177559e+00
50%    -5.233076e-02 -1.245404e-01   3.351431e-01  1.327811e-01
75%     6.722490e-01  5.673506e-01   7.602119e-01  7.879511e-01
max     2.483699e+00  3.104284e+00   1.780377e+00  1.705189e+00
       SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm
count     150.000000    150.000000     150.000000    150.000000
mean        0.428704      0.439167       0.467571      0.457778
std         0.230018      0.180664       0.299054      0.317984
min         0.000000      0.000000       0.000000      0.000000
25%         0.222222      0.333333       0.101695      0.083333
50%         0.416667      0.416667      

#### Penjelasan

Perbedaan hasil normalisasi MinMax dan Z-Score:

1. Rentang nilai: Z-Score berkisar -2.4 hingga 3.1, sementara MinMax selalu 0 sampai 1.

2. Rata-rata: Z-Score mendekati 0 untuk semua fitur, MinMax bervariasi antara 0.42-0.47.

3. Simpangan baku: Z-Score selalu 1, MinMax bervariasi antara 0.18-0.32.

4. Interpretasi: Z-Score menunjukkan jarak nilai dari rata-rata dalam standar deviasi, MinMax menunjukkan posisi relatif nilai dalam rentang aslinya.

5. Sensitivitas terhadap outlier: Z-Score lebih tahan, MinMax lebih sensitif karena bergantung pada nilai ekstrem.

Kedua metode memiliki kelebihan masing-masing tergantung kebutuhan analisis dan karakteristik data. Misalnya, Z-Score lebih tahan terhadap outlier, sedangkan MinMax lebih sensitif karena bergantung pada nilai ekstrem.